### ISSUE TITLE
array copy-on-write issue

### ISSUE THREAD

**Issue by jakubmisek** [peachpiecompiler/peachpie#250]
CoW when modifying nested array is not applied. Test case:
```php
function foo($arr) {
  array_shift( $arr["value"] ); // array_shift gets original $arr["value"], but $arr should have been lazy copied 
}
$arr = ["value" => [1,2,3]];
foo($arr); // foo modifies original "value"
print_r($arr);
```

Expected: [1,2,3]
Actual: [2,3]

**Comment by Ultrafeel** [comment]
Why Have not  anyone integrate tests from https://github.com/php/php-src/tree/master/tests yet?

**Comment by jakubmisek** [comment]
@Ultrafeel that's a good point. Mostly there is a different but expected behavior in edge cases like
- error reporting
- most errors are reported in compile time and test won't run at all
- different `var_dump` behavior (the object ID is different on .NET)
etc.

This makes using existing tests difficult.

For now, let's mark this discussion offtopic and start a new issue for that.

**Comment by jakubmisek** [comment]
update - the fix would affect perf. Needs new impl of PhpArray as a struct instead of a reference type.

**Comment by alecrodden** [comment]
Is resolving this issue on the roadmap? This is definitely a difference with the behavior of regular PHP.

**Comment by jakubmisek** [comment]
update - `array_shift` and other BCL function that modifies its arguments should have parameter attribute (e.g. `[PhpRw]`) so compiler performs copy if necessary (CoW).
